### ISSUE TITLE
Avast's firewall blocks UMS in a strange way

### ISSUE THREAD

**Issue by xeophyte** [UniversalMediaServer/UniversalMediaServer#4787]
UMS is not visible to devices until I turn off the firewall for at least a second.
In the list of Avast firewall rules, both programs allow all network traffic:
- \Universal Media Server\jre17\bin\javaw.exe
- \Universal Media Server\UMS.exe

UMS or OpenJDK Platform binary are not on the blocked list.
All I have to do is turn the firewall off and on again and UMS is visible through the devices, but I have to do it every pc start.
What's wrong?

Avast 24.4.6112
UMS 14.0.2

**Comment by SubJunk** [comment]
Hmm maybe it is blocking some incoming traffic like the broadcast signal from the media player

**Comment by Nadahar** [comment]
It's probably blocking multicast.

**Comment by xeophyte** [comment]
1. My network is set in Avast as Trusted
2. "Allow multicast traffic" in firewall's Basic rules set to enabled at Private networks
3. MULTICAST rule in Network rules set to Private, Allow, In/Out
(There is also "MULTICAST Public" rule, but set to public/block which refers to the public network.)
5. javaw.exe and UMS.exe added to Exceptions and Allowed apps in settings/general.

**Comment by SubJunk** [comment]
As a test, can you try allowing the multicast public rule?

**Comment by Nadahar** [comment]
It's very hard to know what software like Avast choose to block and not in relation to the different options/rules you can make. Only those who wrote the software knows exactly what it does.

The reason I'm pretty convinced it's related to multicast is that in UPnP (that UMS use) only the discovery of other devices depends on multicast. Once the devices have "found" each other, they use unicast. This matches very well with your description of your issue, and it doesn't really matter if multicast is blocked after the discovery is finished, unless of course new devices are connected to the network. New devices won't be seen either, until multicast is allowed again.

I'd say that this is an Avast issue, not UMS issue. It will very likely apply to any other UPnP application you would install too. One thing that *might* be worth trying, is to explicitly allow port 1900 for both UDP and TCP. Try in both directions first, if that works, try one or the other until you figure out what direction is "the problem". I wouldn't hold my breath that this would make a difference though.

**Comment by SubJunk** [comment]
I also looked on our forum and saw that Avast has been the culprit a few times, e.g. https://www.universalmediaserver.com/forum/viewtopic.php?p=48497

**Comment by xeophyte** [comment]
Didn't help. Also, I disabled "Allow multicast traffic" in firewall's Basic rules (so it was disabled for private and public networks) and then I disabled and enabled the firewall (right away) and UMS showed up on the device.
So, I guess it looks like this "Allow multicast traffic" rule has no impact on this problem.

**Comment by xeophyte** [comment]
> As a test, can you try allowing the multicast public rule?

Didn't help.
Also, I disabled "Allow multicast traffic" in firewall's Basic rules (so it was disabled for private and public networks) and then I disabled and enabled the firewall (right away) and UMS showed up on the device.

> I also looked on our forum and saw that Avast has been the culprit a few times, e.g. https://www.universalmediaserver.com/forum/viewtopic.php?p=48497

I don't know if I understood correctly, but if the solution was to switch my network to a trusted in Avast, I have already written that this is how I had it set.

> to explicitly allow port 1900 for both UDP and TCP. Try in both directions first, if that works, try one or the other until you figure out what direction is "the problem"

I set the new rule:
- tcp/udp
- in/out
- local and remote port: 1900
- action: block
and the UMS is showing on the device. 
I don't know if it has to do with the fact that I have already disabled/enabled the firewall or something else. 
I don't understand any of this.

Anyway, maybe I'll try to look for an answer on some Avast forum.

**Comment by Nadahar** [comment]
Allow multicast definitely has to do with it, UMS can't work without multicast. But, it doesn't have to be the only problem.

> I set the new rule:
>
>    tcp/udp
>    in/out
>    local and remote port: 1900
>    action: block
>    and the UMS is showing on the device. I don't know if it has to do with the fact that I have already disabled/enabled the firewall or something else. I don't understand any of this.

`action: block` means "don't allow this traffic", which would not work. My assumption was that the "standard" ports were already opened before having this problem in the first place. Port 1900 should be one of them, I was merely thinking that maybe this rule somehow didn't apply/work so it would be worth testing to open 1900 for any process in any direction.

UMS installs standard "exceptions" to the Windows Firewall rules on installation. Without them it wouldn't work for anyone. The first thing one should do when using a 3. party firewall is to make the corresponding rules in the 3. party firewall. I assume that this is documented somewhere in UMS' documentation.

Being to lazy to search for the documentation, I took a look in UMS' source and discovered that they have changed the "standard" firewall rules to open up MUCH more than they need to. Instead of opening up the specific ports that's needed, they open up *every* port - which is completely unnecessary and I'd say "unsafe". Thus, the source no longer shows which ports needs to be opened. You can see here what rules are created by standard for DMS, a fork of UMS that still only opens what's necessary:

https://github.com/DigitalMediaServer/DigitalMediaServer/blob/a7bc961bcc7bed9a9addc0e6d6309b9aad16a4c2/src/main/external-resources/nsis/setup.nsi#L253-L254

```nsis
nsExec::Exec 'netsh advfirewall firewall add rule name= "Digital Media Server - Incoming TCP port 1900/5252/6363" dir=in action=allow description="Incoming on TCP ports 1900, 5252 and 6363" profile=private,domain protocol=TCP localport=1900,5252,6363 enable=yes'
nsExec::Exec 'netsh advfirewall firewall add rule name= "Digital Media Server - Incoming UDP port 1900" dir=in action=allow description="Incoming on UDP port 1900" profile=private,domain protocol=UDP localport=1900 enable=yes'
```
DMS uses port 5252 for UPnP/DLNA traffic and 6363 for the web interface by default. I don't remember what ports UMS use, but they are not the same. In addition to TCP for these two ports, you need both TCP and UDP for port 1900. All this is incoming access, because Windows firewall allows all outging traffic by default. I'm not sure what outgoing ports would be needed for a firewall that defaults to block outgoing traffic, I don't think that would be viable to specify by port because the other devices on your network can pick whatever port they want to use for their "incoming" traffic, which would need to be allowed out for UMS. Thus, all should be allowed out.

**Comment by SubJunk** [comment]
@xeophyte to clarify the above comment, the UMS Windows firewall rule only allows all ports to be open to UMS, not to any other programs.

Did you find any help on the Avast forums? I'm interested to know if you do so I can add the solution to our knowledge base

**Comment by xeophyte** [comment]
still no replay https://forum.avast.com/index.php?topic=327932.0

**Comment by SubJunk** [comment]
I will install Avast to test it when I’m back home in a couple of days. I use Kaspersky and I don’t have any problems with it

**Comment by Nadahar** [comment]
Anyone using Kaspersky has problems, even though they might not realize it. It should absolutely be avoided: https://en.wikipedia.org/wiki/Kaspersky_bans_and_allegations_of_Russian_government_ties

**Comment by SubJunk** [comment]
I'll avoid getting off topic but I'm satisfied that it's safe. Anyway that's just an example, there are a lot of antivirus and firewall products out there, personally I am mostly guided by the results on [AV Comparatives](https://www.av-comparatives.org/tests/summary-report-2023/)

**Comment by SubJunk** [comment]
@xeophyte I tried Avast and interestingly UMS works for me, tested via my TV and the VLC app for iPhone, but it made my web browser completely stop loading pages, including github.com so I had to disable it to be able to post here. Clearly they have some real bugs with their software. Based on our combined experience it might be good to find a different product. You could also try uninstalling and reinstalling Avast.

The below screenshot is just what it looked like for me when UMS worked. I didn't change any settings from the default, except to enable the Firewall.
![image](https://github.com/user-attachments/assets/25a1585d-128d-46d6-82db-4fc6614a8a5d)

**Comment by xeophyte** [comment]
Hmm, I didn't reinstall Avast yet, but did Repair App and it didn't help, then I did Restore settings and restarted pc and it works now. Then I restored all my Avast settings, but not automatically from a file but manually. And it still works immediately after restarting the PC. I have no idea what changed in the settings, but for now it works immediately without restarting the firewall. 

And this blocking of your browser is strange because after the first launch it should be added to the list of application rules in the firewall and everything should be unblocked.

Anyway, thanks for the tips and I think the topic can be closed.
